apps/aefate/src/aefa_engine_state.erl

```diff
@@ -147,15 +164,48 @@ push_arguments([], Acc, Stack, ES) ->
 push_arguments([A|As], Acc, Stack, ES) ->
     push_arguments(As, A, [Acc | Stack], ES).
 
--spec push_return_address(state()) -> state().
-push_return_address(#es{ current_bb = BB
-                       , current_function = Function
-                       , current_contract = Contract
-                       , call_stack = Stack
-                       , call_value = Value
-                       , memory = Mem} = ES) ->
+-spec push_call_stack(state()) -> state().
+push_call_stack(#es{ current_bb = BB
+                   , current_function = Function
+                   , current_contract = Contract
+                   , call_stack = Stack
+                   , call_value = Value
+                   , memory = Mem} = ES) ->
     ES#es{call_stack = [{Contract, Function, BB+1, Mem, Value}|Stack]}.
 
+%% TODO: Make better types for all these things
+-spec pop_call_stack(state()) ->
+                            'empty' |
+                            {'local', _, non_neg_integer(), state()} |
+                            {'remote', aeb_fate_data:address(), _, non_neg_integer(), state()}.
+pop_call_stack(#es{call_stack = Stack, current_contract = Current} = ES) ->
+    case Stack of
+        [] -> empty;
+        [{gas_store, StoredGas}| Rest] ->
+            ES1 = ES#es{ gas = StoredGas + gas(ES)
+                       , call_stack = Rest
+                       },
+            pop_call_stack(ES1);
+        [{Current, Function, BB, Mem, Value}| Rest] ->
+            {local, Function, BB,
+             ES#es{ call_value = Value
+                  , memory = Mem
+                  , call_stack = Rest
+                  }};
+        [{?FATE_ADDRESS(Pubkey) = Contract, Function, BB, Mem, Value}| Rest] ->
+            Seen = pop_seen_contracts(Pubkey, ES),
+            {remote, Contract, Function, BB,
+             ES#es{ call_value = Value
+                  , memory = Mem
+                  , call_stack = Rest
+                  , seen_contracts = Seen
+                  }}
+    end.
+
+pop_seen_contracts(Pubkey, #es{seen_contracts = Seen}) ->
+    [_|Seen1] = lists:dropwhile(fun(X) -> X =/= Pubkey end, Seen),
```

Comment:
Note: non-`Pubkey` entries are remote tail calls. Maybe a comment would be appropriate.

Reply:
Isn't it obvious? :stuck_out_tongue: 
I'll add a comment on this.